### PR TITLE
ipq40xx-generic: add support for Meraki MR33

### PR DESCRIPTION
**Pull request by maurerle**
This device was tested by someone in our community.
This looks good so far, we are still investigating the open issues and see if we can get support for 5GHz Wifi out of the box too.

- [x] Must be flashable from vendor firmware
  - [ ] Web interface
  - [ ] TFTP
  - [x] Other: https://openwrt.org/toh/meraki/mr33#installation
- [x] Must support upgrade mechanism
  - [x] Must have working sysupgrade
    - [x] Must keep/forget configuration (`sysupgrade [-n]`, `firstboot`)
  - [x] Gluon profile name matches autoupdater image name
        (`lua -e 'print(require("platform_info").get_image_name())'`)
        #output is "meraki-mr33-access-point" - added to manifest_aliases
- [x] Reset/WPS/... button must return device into config mode
- [x] Primary MAC address should match address on device label (or packaging)
      (https://gluon.readthedocs.io/en/latest/dev/hardware.html#hardware-support-in-packages)
      #eth0 MAC matches - set in 010-primary
  - When re-adding a device that was supported by an earlier version of Gluon, a
    factory reset must be performed before checking the primary MAC address, as
    the setting from the old version is not reset otherwise.
- Wired network
  - [x] should support all network ports on the device
  - [x] must have correct port assignment (WAN/LAN)
    - if there are multiple ports but no WAN port:
      - the PoE input should be WAN, all other ports LAN
      - otherwise the first port should be delcared as WAN, all other ports LAN
- Wireless network (if applicable)
  - [x] Association with AP must be possible on all radios
  - [x] Association with 802.11s mesh must work on all radios 
  - [x] AP+mesh mode must work in parallel on all radios
  # monitor radio0 driver is disabled, radio1 is 2.4GHz and radio2 is 5GHz
- LED mapping
  - Power/system LED
    - [x] Lit while the device is on
    - [x] Should display config mode blink sequence 
          (https://gluon.readthedocs.io/en/latest/features/configmode.html)
  - ~~Radio LEDs~~ no radio LEDs
      - ~~Should map to their respective radio~~
      - ~~Should show activity~~
  - Switch port LEDs
    - [x] Should map to their respective port (or switch, if only one led present) 
    - [x] Should show link state and activity
- ~~Outdoor devices only:~~
  - ~~Added board name to `is_outdoor_device` function in `package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua`~~
- ~~Cellular devices only:~~
  - ~~Added board name to `is_cellular_device` function in `package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua`~~
  - ~~Added board name with modem setup function `setup_ncm_qmi` to `package/gluon-core/luasrc/lib/gluon/upgrade/250-cellular`~~
- Docs:
  - ~~Added Device to `docs/user/supported_devices.rst`~~ <- is broken

### CODE REVIEW

**Comment by ambassador86** [comment]
I also have this device in my hands and tried several times to get it running properly. My main issue was the assignment of the third radio device.
The so called monitor radio (which can use both bands) is used as 5GHz interface. The coverage with the monitor radio is not as good as the other two radios. In my firmware the 5GHz only radio was not used at all, but I did not invest enough time to get this running.
See https://github.com/FreifunkChemnitz/gluon/commit/d496c06e0b107e71571939cb14a568969cc4e686

File: targets/ipq40xx-generic
Comment (blocktrron) [inline targets/ipq40xx-generic:103]: Why this alias?

**Review by blocktrron — COMMENTED** [review]
Please add this device as broken, given the board also bricks the SoC on newer bootloader-versions which is irreversible.

**Comment by blocktrron** [comment]
See https://github.com/riptidewave93/LEDE-MR33/issues/13#issuecomment-449849255

File: targets/ipq40xx-generic
Comment (maurerle) [reply targets/ipq40xx-generic:103]: Because it is returned by `lua -e 'print(require("platform_info").get_image_name())'`.
Isn't that whats needed in this case?

**Comment by maurerle** [comment]
@sijans, who was active in this issue a week ago, tested the device. Did not know he is on GitHub :)

Thanks for sharing your experience @ambassador86 !

@sijans can you confirm the flashing procedure which is written in https://github.com/FreifunkChemnitz/gluon/commit/d496c06e0b107e71571939cb14a568969cc4e686 ?

Maybe you can give an update here regarding wifi too?

**Comment by sijans** [comment]
@maurerle Already found me! ;)

The flashing instructions in https://github.com/FreifunkChemnitz/gluon/commit/d496c06e0b107e71571939cb14a568969cc4e686 are missing a few steps. You need to boot a ramdisk image before installing to flash.
I can confirm the [OpenWRT Wiki](https://openwrt.org/toh/meraki/mr33#flash_method_areplace_u-boot) instructions are working, but only if you have the right bootloader version. Yes, no easy way to install gluon. ;)

Regarding wifi, the default gluon configuration looks like this:

```
wireless.radio0=wifi-device
wireless.radio0.type='mac80211'
wireless.radio0.path='soc/40000000.pci/pci0000:00/0000:00:00.0/0000:01:00.0'
wireless.radio0.band='2g'
wireless.radio0.channel='11'
wireless.radio0.htmode='VHT20'
wireless.radio0.country='DE'
wireless.radio0.legacy_rates='0'
wireless.radio1=wifi-device
wireless.radio1.type='mac80211'
wireless.radio1.path='platform/soc/a000000.wifi'
wireless.radio1.band='2g'
wireless.radio1.htmode='HT20'
wireless.radio1.channel='11'
wireless.radio1.country='DE'
wireless.radio1.legacy_rates='0'
wireless.radio2=wifi-device
wireless.radio2.type='mac80211'
wireless.radio2.path='platform/soc/a800000.wifi'
wireless.radio2.band='5g'
wireless.radio2.channel='44'
wireless.radio2.htmode='VHT20'
wireless.radio2.country='DE'
```
Radio0 is not working and floods the log with `kern.warn kernel: [  112.689024] ath10k_pci 0000:01:00.0: failed to send pdev bss chan info request`
Switching radio0 and radio2 by setting
```
wireless.radio0.band='5g'
wireless.radio0.channel='44'
wireless.radio0.path='platform/soc/a800000.wifi'
wireless.radio2.path='soc/40000000.pci/pci0000:00/0000:00:00.0/0000:01:00.0'
```
makes the device work fine with mesh and clients.

**Comment by sijans** [comment]
@maurerle I have attached the requested outputs to this post.
[mr33_ff.txt](https://github.com/freifunk-gluon/gluon/files/11243085/mr33_ff.txt)

**Comment by maurerle** [comment]
Regarding the mac issue:

eth0: `f8:9e:28:71:2e:35` (printed on device)
primary_mac: `f8:9e:28:71:2e:36`

**Does the primary mac need a change?**

The radio0 issue might be fixed using `ath10k-ct` driver - i will build an image for you to test.

Other than that we have to wait for #1661 to support the third radio.
I don't know what the maintainers think of switching the used radio by gluon?

It seems that currently before switching the radio, both client0 and client1 are used in 2.4GHz -
As the first radio - which can do 2.4GHz/5GHz isn't that good as @ambassador86 noted, it might be preferable to use the second and third radio for now (as it is done with @sijans workaround).

**Is there an option to persist the radio switch?**

<details>
<summary>before switching radios:</summary>

```
root@ffac:~# iw dev
phy#1
        Interface client1
                ifindex 18
                wdev 0x100000003
                addr 9e:41:4a:48:ed:6c
                ssid Freifunk
                type AP
                channel 11 (2462 MHz), width: 20 MHz, center1: 2462 MHz
                txpower 20.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       0       0       0       0       0       0               0
        Interface mesh1
                ifindex 16
                wdev 0x100000002
                addr 9e:41:4a:48:ed:6d
                type mesh point
                channel 11 (2462 MHz), width: 20 MHz, center1: 2462 MHz
                txpower 20.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       411     0       0       0       0       32992           411
phy#0
        Interface client0
                ifindex 19
                wdev 0x3
                addr 9e:41:4a:48:ed:68
                ssid Freifunk
                type AP
                channel 11 (2462 MHz), width: 20 MHz, center1: 2462 MHz
                txpower 20.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       0       0       0       0       0       0               0
        Interface mesh0
                ifindex 17
                wdev 0x2
                addr 9e:41:4a:48:ed:69
                type mesh point
                txpower 0.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       0       0       0       0       0       0               0
```

</details>

then running:

```
uci set wireless.radio0.band='5g'
uci set wireless.radio0.channel='44'
uci set wireless.radio0.path='platform/soc/a800000.wifi'
uci set wireless.radio2.path='soc/40000000.pci/pci0000:00/0000:00:00.0/0000:01:00.0'
uci commit wireless
```

<details>
<summary>after switching radios:</summary>

```
root@ffac:~# iw dev
phy#2
        Interface client0
                ifindex 17
                wdev 0x200000003
                addr 9e:41:4a:48:ed:68
                ssid Freifunk
                type AP
                channel 44 (5220 MHz), width: 20 MHz, center1: 5220 MHz
                txpower 23.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       0       0       0       0       0       0               0
        Interface mesh0
                ifindex 15
                wdev 0x200000002
                addr 9e:41:4a:48:ed:69
                type mesh point
                channel 44 (5220 MHz), width: 20 MHz, center1: 5220 MHz
                txpower 23.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       12154   0       0       0       0       1486526         12157
phy#1
        Interface client1
                ifindex 19
                wdev 0x100000003
                addr 9e:41:4a:48:ed:6c
                ssid Freifunk
                type AP
                channel 11 (2462 MHz), width: 20 MHz, center1: 2462 MHz
                txpower 20.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       0       0       0       0       0       0               0
        Interface mesh1
                ifindex 18
                wdev 0x100000002
                addr 9e:41:4a:48:ed:6d
                type mesh point
                channel 11 (2462 MHz), width: 20 MHz, center1: 2462 MHz
                txpower 20.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt hashcol tx-bytes        tx-packets
                        0       0       9640    0       0       0       0       1193860         9641
```

</details>

File: targets/ipq40xx-generic
Comment (blocktrron) [reply targets/ipq40xx-generic:113]: a manifest-alias should suffice for the autoupdater.

Comment (maurerle) [reply targets/ipq40xx-generic]: Right, forgot about `manifest_aliases` parameter - thanks!

**Comment by maurerle** [comment]
I disabled the firmware for radio0 so that gluon takes radio1 (2.4GHz) & radio2 (5GHz).

I also added the MR33 to `010-primary-mac` so that the MAC issue is resolved.
I will get back to it after testing the firmware with this patch

**Comment by sijans** [comment]
I tested the firmware with the patches. Primary MAC matches the label, radio0 is disabled. Gluon now uses radio1 and 2, 2.4 and 5 Ghz are working with client and mesh. This should solve all issues.

File: targets/ipq40xx-generic
Comment (AiyionPrime) [inline targets/ipq40xx-generic:101]: The first reference is effectively matching the string generated from [this line](https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4029-mr33.dts#L7) in OpenWrt.

Why would you use `meraki-mr33` instead here?
I'd suggest changing it here and removing the manifest alias below.


```suggestion
device('meraki-mr33-access-point', 'meraki_mr33', {
```

File: targets/ipq40xx-generic
Comment (AiyionPrime) [inline targets/ipq40xx-generic:110]: Please add _all_ reasons you know of why this device is broken; so that once one is eventually resolved we do not accidentally mark this as supported in the future.

> Please add this device as broken, given the board also bricks the SoC on newer bootloader-versions which is irreversible.

_Originally posted by @blocktrron in https://github.com/freifunk-gluon/gluon/pull/2857#pullrequestreview-1386917707_

File: targets/ipq40xx-generic
Comment (AiyionPrime) [inline targets/ipq40xx-generic:109]:
```suggestion
```

see comment above

**Review by AiyionPrime — CHANGES_REQUESTED** [review]
Please squash your commit `add mr33 to 010-primary-mac` into the one that adds support for the device.

While at it, have a look at the commit contents, in one you are removing `ATH10K_PACKAGES_IPQ40XX_QCA9887` definitions only to add them in the other. Though squashing them might resolve that as well.

File: targets/ipq40xx-generic
Comment (maurerle) [reply targets/ipq40xx-generic]: Oh I didn't knew that it is the name in kebab-case - i thought it was just line 8 with a dash :laughing: 

https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4029-mr33.dts#L8

**Comment by maurerle** [comment]
Thanks @AiyionPrime - it should be ready to be merged now?

**Review by AiyionPrime — APPROVED** [review]
I think so. Maybe @blocktrron can merge this later, if he gets to it.

**Comment by AiyionPrime** [comment]
Just looked over it again,the commit message has a type I'll fix upon merging;
talked to @blocktrron who is fine with the commits contents.

**Pull request merged**
